assignment5/unknownLanguage.py

```diff
@@ -0,0 +1,76 @@
+def unknownLanguage(dictionary):
+    """Given a dictionary of all words in an unknown/invented language, find the
+    alphabet of that language.
+
+    Args:
+        dictionary: a list of words in lexicographic order.
+
+    Returns:
+        the alphabet, an ordered list of characters.
+    """
+    def dfs(v, visited):
+        """
+        Depth-first search with cycle search.
+        """
+        visited[v] = 1
+        for u in adjList[v]:
+            if visited[u] == 0:
+                dfs(u, visited)
+            elif visited[u] == 1:
+                # in that case, we have cycles in our graph =>
+                # it can't be topological sorted => it's not an alphabet
+                raise ValueError("dictionary is not in lexicographic order")
+        visited[v] = 2
+        alphabet.append(v)
+
+    def topologicalSort():
+        """
+        An ordinary topological sort based on dfs.
```

Comment:
Mentions what this operates on and produces. Because it is a local function its arguments and side-effects are not entirely clear. Generally speaking, local functions can introduce a lot more complexity in the code because side-effects are hard to reason about.